linalg: vector norms (#871)
fortran-lang/stdlib: doc/specs/stdlib_linalg.md

{!example/linalg/example_inverse_function.f90!}
```

## `norm` - Computes the vector norm of a generic-rank array.

### Status

Experimental

### Description

This function computes one of several vector norms of `real` or `complex` array \( A \), depending on 
the value of the `order` input argument. \( A \) may be an array of any rank. 

> Result `x` returns a `real`, scalar norm value for the whole array; if `dim` is specified, `x` is a rank-1 
> array with the same shape as \(A \) and dimension `dim` collapsed, containing all norms evaluated along `dim`.

Reviewer: ```suggestion
Result `x` returns a `real` scalar norm value for the whole array; if `dim` is specified, `x` is an array of rank n-1, 
where n equals the rank of ARRAY, and a shape similar to that of \( A \) with dimension `dim` dropped,
containing all norms evaluated along `dim`.
```

|------------------|------------------|---------------------------------------------------------|
| `-huge(0)`       | `'-inf', '-Inf'` | Minimum absolute value \( \min_i{ \left|a_i\right| } \) |
| `1`              | `'1'`            | 1-norm \( \sum_i{ \left|a_i\right| } \)                 |
| `2`              | `'2'`            | Euclidean norm \( \sqrt{\sum_i{ a_i^2 }} \)             |
| `>=3`            | `'3','4',...`    | p-norm \( \left( \sum_i{ \left|a_i\right|^p }\right) ^{1/p} \) |
| `huge(0)`        | `'inf', 'Inf'`   | Maximum absolute value \( \max_i{ \left|a_i\right| } \) |

`dim` (optional): Shall be a scalar `integer` value with a value in the range from `1` to `n`, where `n` is the rank of the array. It is an `intent(in)` argument.

`err` (optional): Shall be a `type(linalg_state_type)` value. This is an `intent(out)` argument. If `err` is not present, the function is `pure`. 

### Return value

By default, the return value `x` is a scalar, and contains the norm as evaluated over all elements of the generic-rank array \( A \). 
If the optional `dim` argument is present, `x` is a rank `n-1` array with the same shape as \( A \) except 

> for dimension `dim`, that is collapsed. Each element of `x` contains the 1D norm of the elements of \( A \), 

Reviewer: ```suggestion
for dimension `dim`, that is dropped. Each element of `x` contains the 1D norm of the elements of \( A \), 
```

evaluated along dimension `dim` only.

Raises `LINALG_ERROR` if the requested norm type is invalid.
Raises `LINALG_VALUE_ERROR` if any of the arguments has an invalid size.
If `err` is not present, exceptions trigger an `error stop`.

### Example

```fortran
{!example/linalg/example_norm.f90!}
```

## `get_norm` - Computes the vector norm of a generic-rank array.

### Status

Experimental

### Description

This `pure subroutine` interface computes one of several vector norms of `real` or `complex` array \( A \), depending on 
the value of the `order` input argument. \( A \) may be an array of any rank. 

> Result `nrm` returns a `real`, scalar norm value for the whole array; if `dim` is specified, `nrm` is a rank-1 

Reviewer: ```suggestion
Result `nrm` returns a `real`, scalar norm value for the whole array; if `dim` is specified, `nrm` is a rank n-1 
```

> array with the same shape as \(A \) and dimension `dim` collapsed, containing all norms evaluated along `dim`.

Reviewer: ```suggestion
array with the same shape as \(A \) and dimension `dim` dropped, containing all norms evaluated along `dim`.
```

{!example/linalg/example_inverse_function.f90!}
```

## `norm` - Computes the vector norm of a generic-rank array.

### Status

Experimental

### Description

This function computes one of several vector norms of `real` or `complex` array \( A \), depending on 
the value of the `order` input argument. \( A \) may be an array of any rank. 

> Result `x` returns a `real` scalar norm value for the whole array; if `dim` is specified, `x` is an array of rank n-1, 

Reviewer 2: This paragraph is mentioned just below in the "return value" section. Maybe it can be dropped from the Description section.

Author: [91882fa](https://github.com/fortran-lang/stdlib/pull/871/commits/91882fa9bf68871f2259033f01a26fff1389b8e4)

| `>=3`            | `'3','4',...`    | p-norm \( \left( \sum_i{ \left|a_i\right|^p }\right) ^{1/p} \) |
| `huge(0)`        | `'inf', 'Inf'`   | Maximum absolute value \( \max_i{ \left|a_i\right| } \) |

`dim` (optional): Shall be a scalar `integer` value with a value in the range from `1` to `n`, where `n` is the rank of the array. It is an `intent(in)` argument.

`err` (optional): Shall be a `type(linalg_state_type)` value. This is an `intent(out)` argument. If `err` is not present, the function is `pure`. 

### Return value

By default, the return value `x` is a scalar, and contains the norm as evaluated over all elements of the generic-rank array \( A \). 
If the optional `dim` argument is present, `x` is a rank `n-1` array with the same shape as \( A \) except 
for dimension `dim`, that is dropped. Each element of `x` contains the 1D norm of the elements of \( A \), 
evaluated along dimension `dim` only.

Raises `LINALG_ERROR` if the requested norm type is invalid.
Raises `LINALG_VALUE_ERROR` if any of the arguments has an invalid size.
If `err` is not present, exceptions trigger an `error stop`.

### Example

```fortran
{!example/linalg/example_norm.f90!}
```

> ## `get_norm` - Computes the vector norm of a generic-rank array.

Reviewer 2: Given that the subroutine contains the core implementation and the function calls it from within, I'm wondering if `get_norm` should be introduced before, and the description of `norm` should make reference to it to also avoid duplicating information and just describing the difference between the funcion interface and the subroutine interface?

Author: [1a79128](https://github.com/fortran-lang/stdlib/pull/871/commits/1a79128d26fd55c103023d5a901fe59bc775b6d9)
